internal/api/validate.go

```diff
@@ -0,0 +1,184 @@
+		// "validate.Var" does not seem like a safe alternative.
+		field := fl.Field()
+		_, kind, nullable := fl.ExtractType(field)
+		switch kind {
+		case reflect.Slice, reflect.Map, reflect.Ptr, reflect.Interface, reflect.Chan, reflect.Func:
+			return !field.IsNil()
+		default:
+			if nullable && field.Interface() != nil {
+				return true
+			}
+			return field.IsValid() && !field.IsZero()
+		}
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	return validate
+}
+
+type validateContext struct {
+	// Fields must be exported so valdator can access.
+	Method   string
+	Resource any
+}
+
+func ValidateRequest(validate *validator.Validate, method string, resource any) []arm.CloudErrorBody {
+	var errorDetails []arm.CloudErrorBody
+
+	err := validate.Struct(validateContext{Method: method, Resource: resource})
+
+	if err == nil {
+		return nil
+	}
+
+	// Convert validation errors to cloud error details.
+	switch err := err.(type) {
+	case validator.ValidationErrors:
+		for _, fieldErr := range err {
+			message := fmt.Sprintf("Invalid value '%s' for field '%s'", fieldErr.Value(), fieldErr.Field())
+			// Try to add a corrective suggestion to the message.
+			tag := fieldErr.Tag()
+			if strings.HasPrefix(tag, "enum_") {
+				message += fmt.Sprintf(" (must be one of: %s)", fieldErr.Param())
+			} else {
+				switch tag {
+				case "api_version": // custom tag
+					message = fmt.Sprintf("Unrecognized API version '%s'", fieldErr.Value())
+				case "required", "required_for_put": // custom tag
+					message = fmt.Sprintf("Missing required field '%s'", fieldErr.Field())
+				case "cidrv4":
+					message += " (must be a v4 CIDR address)"
+				case "ipv4":
+					message += " (must be an IPv4 address)"
+				case "url":
+					message += " (must be a URL)"
+				}
+			}
+			errorDetails = append(errorDetails, arm.CloudErrorBody{
+				Code:    arm.CloudErrorCodeInvalidRequestContent,
+				Message: message,
+				// Split "validateContext.Resource.{REMAINING_FIELDS}"
+				Target: strings.SplitN(fieldErr.Namespace(), ".", 3)[2],
+			})
+		}
+	default:
+		errorDetails = append(errorDetails, arm.CloudErrorBody{
+			Code:    arm.CloudErrorCodeInvalidRequestContent,
+			Message: err.Error(),
+		})
+	}
+
+	return errorDetails
+}
+
+// ValidateSubscription validates a subscription request payload.
+func ValidateSubscription(subscription *arm.Subscription) *arm.CloudError {
+	cloudError := arm.NewCloudError(
+		http.StatusBadRequest,
+		arm.CloudErrorCodeMultipleErrorsOccurred, "",
+		"Content validation failed on multiple fields")
+	cloudError.Details = make([]arm.CloudErrorBody, 0)
+
+	validate := NewValidator()
+	// There is no PATCH method for subscriptions, so assume PUT.
+	errorDetails := ValidateRequest(validate, http.MethodPut, subscription)
+	if errorDetails != nil {
+		cloudError.Details = append(cloudError.Details, errorDetails...)
+	}
+
+	switch len(cloudError.Details) {
+	case 0:
+		cloudError = nil
+	case 1:
```

Comment:
```suggestion
	default:
```
This might be my own ignorance, but is this field guaranteed to have a length of either 0 or 1?

Reply:
The default case (len>1) would leave the error as is, so it's omitted.

In the case of len=1 we rearrange the error structure to avoid ending up with:
```
{
  code: "MultipleErrorsOccurred",
  message: "Content validation failed on multiple fields"
  details: [
    ... only one error is listed, not multiple ...
  ]
}
```

Reply:
Oh, thanks. I wasn't catching that before entering this switch statement, it was already formatted for len>1. This switch statement is solely for modifying the format for len == 0 or 1